### ISSUE TITLE
gevent fails to build on MinGW - libev/ev.c:1767:23: error: storage size of 'ts' isn't known

### ISSUE THREAD

**Issue by jacekt** [gevent/gevent#629]
python-3.4.3-32, Cython 0.22.1, mingw32-gcc 4.8.1-4

```
> type C:\Python34-32\Lib\distutils\distutils.cfg
[build]
compiler = mingw32

> py -3.4-32 setup.py build
Running 'make ' in C:\Sources\3rdParty\gevent
C:\Python34-32\python.exe util/cythonpp.py -o gevent.corecext.c gevent/core.ppyx
Running cython -o gevent.corecext.c gevent/core.pyx  # !EV_USE_SIGNALFD && !defined(LIBEV_EMBED) && !defined(_WIN32)
Running cython -o gevent.corecext.c gevent/core.pyx  # !EV_USE_SIGNALFD && !defined(LIBEV_EMBED) && defined(_WIN32)
Running cython -o gevent.corecext.c gevent/core.pyx  # !EV_USE_SIGNALFD && defined(LIBEV_EMBED) && !defined(_WIN32)
Running cython -o gevent.corecext.c gevent/core.pyx  # !EV_USE_SIGNALFD && defined(LIBEV_EMBED) && defined(_WIN32)
Running cython -o gevent.corecext.c gevent/core.pyx  # EV_USE_SIGNALFD && defined(LIBEV_EMBED) && !defined(_WIN32)
Running cython -o gevent.corecext.c gevent/core.pyx  # EV_USE_SIGNALFD && defined(LIBEV_EMBED) && defined(_WIN32)
Reusing cython -o gevent.corecext.c gevent/core.pyx  # EV_USE_SIGNALFD && defined(LIBEV_EMBED) && !defined(_WIN32)
Reusing cython -o gevent.corecext.c gevent/core.pyx  # EV_USE_SIGNALFD && defined(LIBEV_EMBED) && defined(_WIN32)
Generating gevent.corecext.c 1893733 bytes
Saving gevent/core.pyx
echo '#include "callbacks.c"' >> gevent.corecext.c
mv gevent.corecext.* gevent/
cython -o gevent.ares.c gevent/ares.pyx
mv gevent.ares.* gevent/
cp gevent/_semaphore.pyx gevent/_semaphore.py
cython -o gevent._semaphore.c gevent/_semaphore.py
mv gevent._semaphore.* gevent/
rm gevent/_semaphore.py
cython -o gevent._util.c gevent/_util.pyx
mv gevent._util.* gevent/
running build
running build_py
creating build
creating build\lib.win32-3.4
creating build\lib.win32-3.4\gevent
copying gevent\backdoor.py -> build\lib.win32-3.4\gevent
copying gevent\baseserver.py -> build\lib.win32-3.4\gevent
copying gevent\builtins.py -> build\lib.win32-3.4\gevent
copying gevent\core.py -> build\lib.win32-3.4\gevent
copying gevent\corecffi.py -> build\lib.win32-3.4\gevent
copying gevent\coros.py -> build\lib.win32-3.4\gevent
copying gevent\event.py -> build\lib.win32-3.4\gevent
copying gevent\fileobject.py -> build\lib.win32-3.4\gevent
copying gevent\greenlet.py -> build\lib.win32-3.4\gevent
copying gevent\hub.py -> build\lib.win32-3.4\gevent
copying gevent\local.py -> build\lib.win32-3.4\gevent
copying gevent\lock.py -> build\lib.win32-3.4\gevent
copying gevent\monkey.py -> build\lib.win32-3.4\gevent
copying gevent\os.py -> build\lib.win32-3.4\gevent
copying gevent\pool.py -> build\lib.win32-3.4\gevent
copying gevent\pywsgi.py -> build\lib.win32-3.4\gevent
copying gevent\queue.py -> build\lib.win32-3.4\gevent
copying gevent\resolver_ares.py -> build\lib.win32-3.4\gevent
copying gevent\resolver_thread.py -> build\lib.win32-3.4\gevent
copying gevent\select.py -> build\lib.win32-3.4\gevent
copying gevent\server.py -> build\lib.win32-3.4\gevent
copying gevent\socket.py -> build\lib.win32-3.4\gevent
copying gevent\ssl.py -> build\lib.win32-3.4\gevent
copying gevent\subprocess.py -> build\lib.win32-3.4\gevent
copying gevent\thread.py -> build\lib.win32-3.4\gevent
copying gevent\threading.py -> build\lib.win32-3.4\gevent
copying gevent\threadpool.py -> build\lib.win32-3.4\gevent
copying gevent\timeout.py -> build\lib.win32-3.4\gevent
copying gevent\util.py -> build\lib.win32-3.4\gevent
copying gevent\win32util.py -> build\lib.win32-3.4\gevent
copying gevent\wsgi.py -> build\lib.win32-3.4\gevent
copying gevent\_fileobject2.py -> build\lib.win32-3.4\gevent
copying gevent\_fileobject3.py -> build\lib.win32-3.4\gevent
copying gevent\_fileobjectcommon.py -> build\lib.win32-3.4\gevent
copying gevent\_fileobjectposix.py -> build\lib.win32-3.4\gevent
copying gevent\_socket2.py -> build\lib.win32-3.4\gevent
copying gevent\_socket3.py -> build\lib.win32-3.4\gevent
copying gevent\_socketcommon.py -> build\lib.win32-3.4\gevent
copying gevent\_ssl2.py -> build\lib.win32-3.4\gevent
copying gevent\_ssl3.py -> build\lib.win32-3.4\gevent
copying gevent\_sslgte279.py -> build\lib.win32-3.4\gevent
copying gevent\_tblib.py -> build\lib.win32-3.4\gevent
copying gevent\_threading.py -> build\lib.win32-3.4\gevent
copying gevent\_util_py2.py -> build\lib.win32-3.4\gevent
copying gevent\__init__.py -> build\lib.win32-3.4\gevent
running build_ext
building 'gevent.core' extension
creating build\temp.win32-3.4
creating build\temp.win32-3.4\Release
creating build\temp.win32-3.4\Release\gevent
c:\MinGW\bin\gcc.exe -mdll -O -Wall -DFD_SETSIZE=1024 -D_WIN32=1 -DLIBEV_EMBED=1 -DEV_COMMON= -DEV_CLEANUP_ENABLE=0 -DEV_EMBED_ENABLE=0 -DEV_PERIODIC_ENABLE=0 -
DCARES_STATICLIB= -DCARES_EMBED=1 -DEV_STANDALONE=1 -Ilibev -IC:\Python34-32\include -IC:\Python34-32\include -c gevent/gevent.corecext.c -o build\temp.win32-3.
4\Release\gevent\gevent.corecext.o
In file included from gevent/libev.h:2:0,
                 from gevent/gevent.corecext.c:256:
libev/ev.c:483:48: warning: "/*" within comment [-Wcomment]
 /*#define MIN_INTERVAL  0.00000095367431640625 /* 1/2**20, good till 2200 */
 ^
In file included from gevent/libev.h:2:0,
                 from gevent/gevent.corecext.c:256:
libev/ev.c:1692:31: warning: 'ev_default_loop_ptr' initialized and declared 'extern' [enabled by default]
   EV_API_DECL struct ev_loop *ev_default_loop_ptr = 0; /* needs to be initialised to make it a definition despite extern */
                               ^
libev/ev.c: In function 'ev_sleep':
libev/ev.c:1767:23: error: storage size of 'ts' isn't known
       struct timespec ts;
                       ^
libev/ev.c:1770:7: warning: implicit declaration of function 'nanosleep' [-Wimplicit-function-declaration]
       nanosleep (&ts, 0);
       ^
libev/ev.c:1767:23: warning: unused variable 'ts' [-Wunused-variable]
       struct timespec ts;
                       ^
In file included from libev/ev.c:174:0,
                 from gevent/libev.h:2,
                 from gevent/gevent.corecext.c:256:
libev/ev.c: In function 'fd_reify':
libev/ev.c:1951:80: warning: left-hand operand of comma expression has no effect [-Wunused-value]
               assert (("libev: only socket fds supported in this configuration", ioctlsocket (handle, FIONREAD, &arg) == 0));
                                                                                ^
In file included from gevent/libev.h:2:0,
                 from gevent/gevent.corecext.c:256:
libev/ev.c: In function 'evpipe_write':
libev/ev.c:2329:19: warning: assignment from incompatible pointer type [enabled by default]
           buf.buf = &buf;
                   ^
In file included from libev/ev.c:174:0,
                 from gevent/libev.h:2,
                 from gevent/gevent.corecext.c:256:
libev/ev_select.c: In function 'select_modify':
libev/ev_select.c:85:77: warning: left-hand operand of comma expression has no effect [-Wunused-value]
     assert (("libev: fd >= FD_SETSIZE passed to fd_set-based select backend", fd < FD_SETSIZE));
                                                                             ^
In file included from libev/ev.c:2560:0,
                 from gevent/libev.h:2,
                 from gevent/gevent.corecext.c:256:
libev/ev_select.c:91:8: warning: suggest explicit braces to avoid ambiguous 'else' [-Wparentheses]
     if ((oev ^ nev) & EV_READ)
        ^
libev/ev_select.c:99:8: warning: suggest explicit braces to avoid ambiguous 'else' [-Wparentheses]
     if ((oev ^ nev) & EV_WRITE)
        ^
libev/ev_select.c: In function 'select_poll':
libev/ev_select.c:181:10: warning: suggest explicit braces to avoid ambiguous 'else' [-Wparentheses]
       if (errno >= WSABASEERR && errno < WSABASEERR + 1000)
          ^
In file included from libev/ev.c:174:0,
                 from gevent/libev.h:2,
                 from gevent/gevent.corecext.c:256:
libev/ev.c: In function 'verify_watcher':
libev/ev.c:2963:49: warning: left-hand operand of comma expression has no effect [-Wunused-value]
   assert (("libev: watcher has invalid priority", ABSPRI (w) >= 0 && ABSPRI (w) < NUMPRI));
                                                 ^
libev/ev.c:2966:59: warning: left-hand operand of comma expression has no effect [-Wunused-value]
     assert (("libev: pending watcher not on pending queue", pendings [ABSPRI (w)][w->pending - 1].w == w));
                                                           ^
libev/ev.c: In function 'verify_heap':
libev/ev.c:2976:54: warning: left-hand operand of comma expression has no effect [-Wunused-value]
       assert (("libev: active index mismatch in heap", ev_active (ANHE_w (heap [i])) == i));
                                                      ^
libev/ev.c:2977:48: warning: left-hand operand of comma expression has no effect [-Wunused-value]
       assert (("libev: heap condition violated", i == HEAP0 || ANHE_at (heap [HPARENT (i)]) <= ANHE_at (heap [i])));
                                                ^
libev/ev.c:2978:47: warning: left-hand operand of comma expression has no effect [-Wunused-value]
       assert (("libev: heap at cache mismatch", ANHE_at (heap [i]) == ev_at (ANHE_w (heap [i]))));
                                               ^
libev/ev.c: In function 'array_verify':
libev/ev.c:2989:46: warning: left-hand operand of comma expression has no effect [-Wunused-value]
       assert (("libev: active index mismatch", ev_active (ws [cnt]) == cnt + 1));
                                              ^
libev/ev.c: In function 'ev_verify':
libev/ev.c:3007:47: warning: left-hand operand of comma expression has no effect [-Wunused-value]
     assert (("libev: negative fd in fdchanges", fdchanges [i] >= 0));
                                               ^
libev/ev.c:3020:64: warning: left-hand operand of comma expression has no effect [-Wunused-value]
               assert (("libev: io watcher list contains a loop", w != w2));
                                                                ^
libev/ev.c:3024:61: warning: left-hand operand of comma expression has no effect [-Wunused-value]
           assert (("libev: inactive fd watcher on anfd list", ev_active (w) == 1));
                                                             ^
libev/ev.c:3025:65: warning: left-hand operand of comma expression has no effect [-Wunused-value]
           assert (("libev: fd mismatch between watcher and anfd", ((ev_io *)w)->fd == i));
                                                                 ^
libev/ev.c: In function 'timers_reify':
libev/ev.c:3207:93: warning: left-hand operand of comma expression has no effect [-Wunused-value]
               assert (("libev: negative ev_timer repeat value found while processing timers", w->repeat > 0.));
                                                                                             ^
libev/ev.c: In function 'ev_run':
libev/ev.c:3408:62: warning: left-hand operand of comma expression has no effect [-Wunused-value]
   assert (("libev: ev_loop recursion during release detected", loop_done != EVBREAK_RECURSE));
                                                              ^
libev/ev.c:3529:70: warning: left-hand operand of comma expression has no effect [-Wunused-value]
             assert (("libev: pipe_w not active, but pipe not written", ev_is_active (&pipe_w)));
                                                                      ^
libev/ev.c: In function 'ev_io_start':
libev/ev.c:3703:56: warning: left-hand operand of comma expression has no effect [-Wunused-value]
   assert (("libev: ev_io_start called with negative fd", fd >= 0));
                                                        ^
libev/ev.c:3704:63: warning: left-hand operand of comma expression has no effect [-Wunused-value]
   assert (("libev: ev_io_start called with illegal event mask", !(w->events & ~(EV__IOFDSET | EV_READ | EV_WRITE))));
                                                               ^
libev/ev.c:3713:62: warning: left-hand operand of comma expression has no effect [-Wunused-value]
   assert (("libev: ev_io_start called with corrupted watcher", ((WL)w)->next != (WL)w));
                                                              ^
In file included from gevent/libev.h:2:0,
                 from gevent/gevent.corecext.c:256:
libev/ev.c:3715:34: warning: suggest parentheses around arithmetic in operand of '|' [-Wparentheses]
   fd_change (EV_A_ fd, w->events & EV__IOFDSET | EV_ANFD_REIFY);
                                  ^
In file included from libev/ev.c:174:0,
                 from gevent/libev.h:2,
                 from gevent/gevent.corecext.c:256:
libev/ev.c: In function 'ev_io_stop':
libev/ev.c:3728:88: warning: left-hand operand of comma expression has no effect [-Wunused-value]
   assert (("libev: ev_io_stop called with illegal fd (must stay constant after start!)", w->fd >= 0 && w->fd < anfdmax));
                                                                                        ^
libev/ev.c: In function 'ev_timer_start':
libev/ev.c:3748:75: warning: left-hand operand of comma expression has no effect [-Wunused-value]
   assert (("libev: ev_timer_start called with negative timer repeat value", w->repeat >= 0.));
                                                                           ^
libev/ev.c: In function 'ev_timer_stop':
libev/ev.c:3776:53: warning: left-hand operand of comma expression has no effect [-Wunused-value]
     assert (("libev: internal timer heap corruption", ANHE_w (timers [active]) == (WT)w));
                                                     ^
libev/ev.c: In function 'ev_signal_start':
libev/ev.c:3907:70: warning: left-hand operand of comma expression has no effect [-Wunused-value]
   assert (("libev: ev_signal_start called with illegal signal number", w->signum > 0 && w->signum < EV_NSIG));
                                                                      ^
libev/ev.c:3910:73: warning: left-hand operand of comma expression has no effect [-Wunused-value]
   assert (("libev: a signal must not be attached to two different loops",
                                                                         ^
In file included from gevent/libev.h:2:0,
                 from gevent/gevent.corecext.c:256:
libev/ev.c: At top level:
libev/ev.c:4956:27: warning: "/*" within comment [-Wcomment]
 /* EV_STAT     0x00001000 /* stat data changed */
 ^
libev/ev.c:4957:27: warning: "/*" within comment [-Wcomment]
 /* EV_EMBED    0x00010000 /* embedded event loop needs sweep */
 ^
gevent/gevent.corecext.c: In function '__pyx_pf_6gevent_4core_4stat_4attr___get__':
gevent/gevent.corecext.c:33907:3: warning: passing argument 1 of '_pystat_fromstructstat' from incompatible pointer type [enabled by default]
   __pyx_t_2 = _pystat_fromstructstat((&__pyx_v_self->_watcher.attr)); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 2046; __pyx_clinen
o = __LINE__; goto __pyx_L1_error;}
   ^
In file included from gevent/gevent.corecext.c:259:0:
gevent/stathelper.c:89:1: note: expected 'struct stat *' but argument is of type 'struct ev_statdata *'
 _pystat_fromstructstat(STRUCT_STAT *st)
 ^
gevent/gevent.corecext.c: In function '__pyx_pf_6gevent_4core_4stat_4prev___get__':
gevent/gevent.corecext.c:33993:3: warning: passing argument 1 of '_pystat_fromstructstat' from incompatible pointer type [enabled by default]
   __pyx_t_2 = _pystat_fromstructstat((&__pyx_v_self->_watcher.prev)); if (unlikely(!__pyx_t_2)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 2053; __pyx_clinen
o = __LINE__; goto __pyx_L1_error;}
   ^
In file included from gevent/gevent.corecext.c:259:0:
gevent/stathelper.c:89:1: note: expected 'struct stat *' but argument is of type 'struct ev_statdata *'
 _pystat_fromstructstat(STRUCT_STAT *st)
 ^
In file included from gevent/libev.h:2:0,
                 from gevent/gevent.corecext.c:256:
gevent/gevent.corecext.c: At top level:
libev/ev.c:1457:1: warning: 'ev_floor' defined but not used [-Wunused-function]
 ev_floor (ev_tstamp v)
 ^
In file included from gevent/libev.h:2:0,
                 from gevent/gevent.corecext.c:256:
libev/ev.c:2058:1: warning: 'fd_rearm_all' defined but not used [-Wunused-function]
 fd_rearm_all (EV_P)
 ^
In file included from gevent/gevent.corecext.c:256:0:
gevent/libev.h:44:13: warning: 'gevent_install_sigchld_handler' defined but not used [-Wunused-function]
 static void gevent_install_sigchld_handler(void) { }
             ^
error: command 'c:\\MinGW\\bin\\gcc.exe' failed with exit status 1
```

**Comment by parasyte** [comment]
This appears to be a bug in MinGW 3.21: http://thread.gmane.org/gmane.comp.gnu.mingw.user/44416/focus=44420 It's supposed to be fixed in 3.21.1: http://sourceforge.net/p/mingw/mingw-org-wsl/ci/mingwrt-3.21.1-release/tree/mingwrt/ChangeLog

**Comment by jamadden** [comment]
Closing per the previous comment. Please feel free to comment or reopen if this doesn't seem to be the problem.
